gemma-core/src/main/java/ubic/gemma/core/ontology/OntologyServiceImpl.java

```diff
@@ -635,58 +635,60 @@ public Collection<Characteristic> termsToCharacteristics( final Collection<? ext
     }
 
     @Override
-    public Map<String, CharacteristicValueObject> countObsoleteOccurrences( int start, int stop, int step ) {
+    public Map<String, CharacteristicValueObject> findObsoleteTermUsage() {
         Map<String, CharacteristicValueObject> vos = new HashMap<>();
 
-        int minId = start;
-        int maxId = step;
+        int start = 0;
+        int step = 5000;
 
-        int nullCnt = 0;
-        int obsoleteCnt = 0;
+        int prevObsoleteCnt = 0;
+        int checked = 0;
+        CharacteristicValueObject lastObsolete = null;
 
-        // Loading all characteristics in steps
-        while ( maxId < stop ) {
+        while ( true ) {
 
-            OntologyServiceImpl.log.info( "Checking characteristics with IDs between " + minId + " and " + maxId );
+            Collection<Characteristic> chars = characteristicService.browse( start, step );
+            start += step;
 
-            List<Long> ids = new ArrayList<>( step );
-            for ( int i = minId; i < maxId + 1; i++ ) {
-                ids.add( ( long ) i );
+            if ( chars == null || chars.isEmpty() ) {
+                break;
             }
 
-            minId = maxId + 1;
-            maxId += step;
+            for ( Characteristic ch : chars ) {
+                String valueUri = ch.getValueUri();
+                if ( StringUtils.isBlank( valueUri ) ) {
+                    continue;
+                }
 
-            Collection<Characteristic> chars = characteristicService.load( ids );
+                checked++;
 
-            if ( chars == null || chars.isEmpty() ) {
-                OntologyServiceImpl.log.info( "No characteristics in the current ID range, moving on." );
-                continue;
-            }
-            OntologyServiceImpl.log.info( "Found " + chars.size()
-                    + " characteristics in the current ID range, checking for obsoletes." );
+                if ( this.getTerm( valueUri ) == null || this.isObsolete( valueUri ) ) {
```

Comment:
It is more efficient to call `isObsolete` on the term itself.

Reply:
You should also check if the term was already marked previously to avoid hitting the ontology service repeatedly with the same URIs.

Reply:
I might make a `getTerms(Collection)` since that would better use the threadpool.

Reply:
That makes sense as long as we still enter the block: we need to increment the count. Having a count of usages is important to prioritize curation.